Lwm2m timeseries update

jheiskan81 opened the pull request:
**Read by server update**

 Normal Read by server only report back latest stored data.

**Notification and Send operation update**

Added support for handle case when all data is not possible to add in 1 message for Send and Observed Notification. Engine rebuild message by limiting time series cache data length. 

Notification continuous pending time series data is triggered by iMIN attribute. Send Operation generate continuous message in multiple lwm2m message.

Commit:
net: lwm2m: Timeseries data cache read update

Added support for handle case when all data is not possible to
add in 1 message for Send and Observed Notification.

Notification continuous pending timeseries data is triggred
by iMIN attribute.

Send Operation generate continuous message in multiple lwm2m
message.

Normal Read by server only report back latest stored data.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@SeppoTakalo I dropped first commit away.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_registry.h:
```diff
@@ -212,6 +212,27 @@ struct lwm2m_time_series_resource {
 	struct ring_buf rb;
 };
 
+
+
+
```
No need for those emtpy lines.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -315,6 +315,9 @@ void lwm2m_reset_message(struct lwm2m_message *msg, bool release)
 	} else {
 		msg->message_timeout_cb = NULL;
 		(void)memset(&msg->cpkt, 0, sizeof(msg->cpkt));
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
```
It just struck me, we should add `CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT` to `tests/net/all/prj.conf`, otherwise this variant won't be built by the CI (easy to introduce regressions).

jheiskan81 replied:
OK. How we manage dependencies
_depends on (LWM2M_RW_SENML_JSON_SUPPORT || LWM2M_RW_SENML_CBOR_SUPPORT)
depends on (POSIX_CLOCK && RING_BUFFER):_

jheiskan81 replied:
Or was a idea set CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT=n?

jheiskan81 replied:
I added CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT=y at PR it not triggered any issue. @rlubos Is that correct implementation about what did you request?

rlubos replied:
yes, that's what I had in mind.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2549,13 +2652,19 @@ int generate_notify_message(struct lwm2m_ctx *ctx, struct observe_node *obs, voi
 	}
 
 	if (ret < 0) {
+		if (lwm2m_timeseries_data_rebuild(msg, ret)) {
+			goto msg_init;
+		}
```
I think some comment explaining logic here could be useful, it's not clear to me what is the retry condition.

jheiskan81 replied:
Ok I will add comment's for this.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2600,6 +2709,7 @@ int lwm2m_perform_composite_read_op(struct lwm2m_message *msg, uint16_t content_
 {
 	struct lwm2m_engine_obj_inst *obj_inst = NULL;
 	struct lwm2m_obj_path_list *entry;
+
```
Unneeded empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2830,6 +2940,47 @@ static void do_send_timeout_cb(struct lwm2m_message *msg)
+					  sys_slist_t *lwm2m_path_list,
+					  sys_slist_t *lwm2m_path_free_list)
+{
+
```
Unneeded empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2926,12 +3091,33 @@ int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t pa
+
+		if (init_next_pending_timeseries_data(&cache_temp_info, &lwm2m_path_list,
+						     &lwm2m_path_free_list)) {
+			goto msg_alloc;
```
Same story here, this goto logic could use some explaination.

jheiskan81 replied:
Ok . I will add comment's.

jheiskan81 commented:
Rebased with upstream

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2472,18 +2522,68 @@ static int do_send_op(struct lwm2m_message *msg, uint16_t content_format,
+	}
+
+	/* Put Ring buffer back to original */
+	for (int i = 0; i < cache_temp->entry_size; i++) {
+		cache_temp->read_info[i].cache_data->rb.get_head =
+			cache_temp->read_info[i].original_get_head;
+		cache_temp->read_info[i].cache_data->rb.get_tail =
+			cache_temp->read_info[i].original_get_tail;
+		cache_temp->read_info[i].cache_data->rb.get_base =
+			cache_temp->read_info[i].original_get_base;
+	}
```
Do we have the ring buffer locked while we do this?
Otherwise the head might go forward.

jheiskan81 replied:
Good Point. I need to change lwm2m_engine_send() that whole process is secured.

Commit:
net: lwm2m: Send operation registry lock update

Secured Message allocation, init and rebuild process for secure
timeseries ring buffer.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@SeppoTakalo I fixed Ring buffer secure. Send operation was not secured correctly.
